File: cpp_merge/packages/integration/SWEagent/src/agent/hooks/status.cpp
#include "status.hpp"

SetStatusAgentHook::SetStatusAgentHook(string id, std::function<void(string, string)> callable) : AbstractAgentHook() {
    this->id = id;
    this->callable = callable;
}

void SetStatusAgentHook::onSetupAttempt()
{
    this->iAttempt++;
    this->iStep = 0;
    this->previousCost += this->cost;
    this->cost = 0;
}

void SetStatusAgentHook::update(string message)
{
    this->callable(this->id, message);
}

void SetStatusAgentHook::onStepStart()
{
    this->iStep++;
    auto attemptStr = (this->iAttempt > 1) ? any(std:("Attempt ") + this->iAttempt + std:(" ")) (string_empty);
    this->update(string_empty + attemptStr + std:("Step ") + this->iStep->toString()->padStart(3) + std:(" ($") + (this->previousCost + this->cost)->toFixed(2) + std:(")"));
}

void SetStatusAgentHook::onStepDone(std::shared_ptr<StepOutput> _step, std::shared_ptr<AgentInfo> info)
{
    if (AND((info->modelStats), (type_of(info->modelStats->instanceCost) == std:("number")))) {
        this->cost = info->modelStats->instanceCost;
    }
}

void SetStatusAgentHook::onToolsInstallationStarted()
{
    this->update(std:("Installing tools"));
}

File: SWEagent/src/agent/hooks/status.ts
/**
 * Status hook for updating agent status
 * Converted from sweagent/agent/hooks/status.py
 */

import { AbstractAgentHook } from './abstract';
import { AgentInfo, StepOutput } from '../../types';

/**
 * Hook for setting status during agent execution
 */
export class SetStatusAgentHook extends AbstractAgentHook {
  private callable: (id: string, message: string) => void;
  private id: string;
  private iStep: number = 0;
  private cost: number = 0.0;
  private iAttempt: number = 0;
  private previousCost: number = 0.0;

  constructor(id: string, callable: (id: string, message: string) => void) {
    super();
    this.id = id;
    this.callable = callable;
  }

  onSetupAttempt(): void {
    this.iAttempt++;
    this.iStep = 0;
    // Costs will be reset for the next attempt
    this.previousCost += this.cost;
    this.cost = 0.0;
  }

  private update(message: string): void {
    this.callable(this.id, message);
  }

  onStepStart(): void {
    this.iStep++;
    const attemptStr = this.iAttempt > 1 ? `Attempt ${this.iAttempt} ` : '';
    this.update(
      `${attemptStr}Step ${this.iStep.toString().padStart(3)} ($${(this.previousCost + this.cost).toFixed(2)})`,
    );
  }

  onStepDone(_step: StepOutput, info: AgentInfo): void {
    // Update cost from model stats
    if (info.modelStats && typeof info.modelStats.instanceCost === 'number') {
      this.cost = info.modelStats.instanceCost;
    }
  }

  onToolsInstallationStarted(): void {
    this.update('Installing tools');
  }
}
